text/0000-js-compatible-resolution.md

```diff
@@ -0,0 +1,344 @@
+---
+
+<Select />
+<MultiSelect />
+```
+
+We require component Javascript to be _authored_ using default exports (because we associate templates with default exports), but the resulting components can still be grouped and shared via named exports.
+
+Another nice feature that necessarily works in this model is that helpers can be named exports grouped into one file:
+
+```js
+import { helper } from '@ember/component/helper';
+
+const gt = helper(function([left, right])  {
+  return left > right;
+}
+
+const lt = helper(function([left, right]) {
+  return left < right;
+}
+
+export { gt, lt };
+```
+
+```hbs
+---
+import { gt, lt } from 'the-helpers-module';
+---
+{{#if (gt this.thing 10)}}
+  more than 10
+{{/if}}
+
+{{#if (lt this.thing 1)}}
+  less than 1
+{{/if}}
+```
+
+In fact, helpers can colocate with components in one file if you want them to:
+
+```js
+// my-component/index.js
+import Component from '@ember/component';
+import { helper } from '@ember/component/helper';
+
+export default class extends Component({
+
+});
+
+export const customHelper = helper(function(){
+  return 'hello world';
+})
+```
+
+```hbs
+{{! my-component/index.hbs }}
+---
+import { customHelper } from '.';
+---
+<div class={{customHelper}} />
+```
+
+## App structure implications
+
+Most of the MU structure remains unchanged. Specifically, you still need things like `src/data/models`, `src/init`, `/src/services`, and `/src/ui/routes/` because all these things are still handled by the Collection Resolver.
+
+But _within_ component collections (meaning `src/ui/components` plus any `-components` directories nested under routes) we don't need to impose any particular structure. You can nest components where they're used, and access them via relative imports.
+
+Within component collections, there is no longer any need for special `-components` or `-utils` private collections. And the names `component.js` and `template.hbs` lose their special meaning. There are no special rules for looking up components that differ from looking up any other modules.
+
+## Addon structure implications
+
+An addon authored in MU is relatively free to organize its components and helpers as it wishes within the `/src/ui/components` directory. Addon authors should provide a top-level entrypoint that re-exports public components & helpers so they're accessible without deep specifiers, and so it's clear what consitutes the addon's public API.
+
+The prior MU RFCs were vague on top-level entrypoints, but they were consistent in insisting that import specifiers should equal on-disk paths. In concordance with that, we clarify that the top-level entrypoint of an MU addon (one that contains a `/src` dir) is its true `index.js` at the addon's root. The ember-cli-specific code that often appears there instead can be placed elsewhere thanks to the `ember-addon.main` option in package.json. (This is a preexisting feature, not a new feature in this RFC.)
```

Comment:
We should not use the addon's top level `index.js` by default (though I totally agree that it should "work", it just shouldn't be what we recommend). Instead we should specify `main` as `src/index.js` _and_ `ember-addon.main` (as you suggest below).

Specifically, this is due to an issue in ESLint. It is not possible to target rules at `index.js` in the project root without _also_ applying those rules to **all** `index.js` files throughout the entire project. Using `src/index.js` + `"main": "src/index.js",` avoids this issue and still nicely follows _very very_ common ecosystem patterns.